src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1012 @@
+	 * this case the carryover is a valid replica for the next stripe)
+	 */
+	}
+	return 0;
+}
+
+/* Clears the extent list of all extents completed for the processed stripe.
+ * Extents the carry over to the next stripe have the prior-stripe prefix
+ * trimmed.
+ */
+static void
+agg_clear_extents(struct ec_agg_entry *agg_entry)
+{
+	struct ec_agg_extent *agg_extent, *ext_tmp;
+	unsigned int	      tail, ptail = 0U;
+
+	agg_entry->ae_cur_stripe.as_prefix_ext = 0U;
+	d_list_for_each_entry_safe(agg_extent, ext_tmp,
+				   &agg_entry->ae_cur_stripe.as_dextents,
+				   ae_link) {
+		/* Check for carry-over extent. */
+		tail = agg_carry_over(agg_entry, agg_extent);
+		/* At most one extent should carry over.( */
+		if (tail) {
+			D_ASSERT(ptail == 0U);
+			ptail = tail;
+			agg_entry->ae_cur_stripe.as_prefix_ext =
+					agg_extent->ae_recx.rx_nr - tail;
+			agg_extent->ae_recx.rx_idx +=
+					agg_extent->ae_recx.rx_nr - tail;
+			agg_extent->ae_recx.rx_nr = tail;
+			agg_entry->ae_cur_stripe.as_hi_epoch =
+				agg_extent->ae_epoch;
+		} else {
+			agg_entry->ae_cur_stripe.as_extent_cnt--;
+			d_list_del(&agg_extent->ae_link);
+			D_FREE_PTR(agg_extent);
+		}
+	}
+	agg_entry->ae_cur_stripe.as_offset = 0U;
+	/* Account for carry over. */
+	if (ptail)
+		agg_entry->ae_cur_stripe.as_stripenum++;
+	else
+		agg_entry->ae_cur_stripe.as_hi_epoch = 0UL;
+	agg_entry->ae_cur_stripe.as_stripe_fill = ptail;
+}
+
+/* Returns the stripe number for the stripe containing ex_lo.
+ */
+static inline daos_off_t
+agg_stripenum(struct ec_agg_entry *entry, daos_off_t ex_lo)
+{
+	return ex_lo / obj_ec_stripe_rec_nr(entry->ae_oca);
+}
+
+/* Call back for the nested iterator used to find the parity for a stripe.
+ */
+static int
+agg_recx_iter_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		     vos_iter_type_t type, vos_iter_param_t *param,
+		     void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *)cb_arg;
+	int			 rc = 0;
+
+	D_ASSERT(type == VOS_ITER_RECX);
+	D_ASSERT(entry->ie_recx.rx_idx == (PARITY_INDICATOR |
+		(agg_entry->ae_cur_stripe.as_stripenum *
+			agg_entry->ae_oca->u.ec.e_len)));
+	agg_entry->ae_par_extent.ape_recx = entry->ie_recx;
+	agg_entry->ae_par_extent.ape_epoch = entry->ie_epoch;
+	return rc;
+}
+
+enum agg_iov_entry {
+	AGG_IOV_DATA	= 0,
+	AGG_IOV_ODATA,
+	AGG_IOV_PARITY,
+	AGG_IOV_DIFF,
+	AGG_IOV_CNT,
+};
+
+/* Allocates an sgl iov_buf at iov_entry offset in the array.
+ */
+static int
+agg_alloc_buf(d_sg_list_t *sgl, size_t ent_buf_len, unsigned int iov_entry,
+	      bool align_data)
+{
+	int		 rc = 0;
+
+	if (align_data) {
+		D_FREE(sgl->sg_iovs[iov_entry].iov_buf);
+		sgl->sg_iovs[iov_entry].iov_buf =
+			aligned_alloc(32, ent_buf_len);
+		if (sgl->sg_iovs[iov_entry].iov_buf == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+	} else {
+		unsigned int *buf = NULL;
+
+		D_REALLOC(buf, sgl->sg_iovs[iov_entry].iov_buf, ent_buf_len);
```

Comment:
only realloc if ent_buf_len > sgl->sg_iovs[].iov_buf_len?

Reply:
The check is performed prior to calling agg_alloc_buf().